### PR TITLE
Name resolution

### PR DESCRIPTION
**Pull request by woocash2**

### CODE REVIEW

File: app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt [hunk 1]
```diff
@@ -1,12 +1,269 @@
...
+                    analyzeNode(node.variable, currentScope)
+                    addName(node.variable.name, node.variable, currentScope)
+                }
+
+                is Statement.FunctionDefinition -> {
+                    reportIfNameConflict(node.function.name, currentScope)
+
+                    // first add name, then analyze because we can have recursive calls in the body
+                    addName(node.function.name, node.function, currentScope)
+                    analyzeNode(node.function, currentScope)
```
Comment (kamgal00) [inline app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt:224]: Cases `Statement.FunctionDefinition` and `Global.FunctionDefinition` are basically the same - could we move the repeated code to the `Function` case and here simply write `is Statement.FunctionDefinition -> analyzeNode(node.function, currentScope)`? Could we do the same for `Statement.VariableDefinition`/`Global.VariableDefinition`?

**Review by kamgal00 — APPROVED** [review]
Kod jest czytelny, zgrabnie napisany i dobrze opisany komentarzami :) dodałem jedną małą uwagę, tak to chyba wszystko git

File: app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt [hunk 2]
```diff
@@ -1,12 +1,269 @@
...
+                }
+
+                is Statement.Assignment -> {
+                    reportIfVariableUndefined(node.variableName)
+                    reportIfFunctionUsedAsVariable(node.variableName) // when we try to assign to a function
+                    nameDefinitions[node.variableName] = visibleNames[node.variableName]!!.topVariable()
```
Comment (rkilar) [inline app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt:230]: Shouldn't the key be be `node` not `node.variableName`?

File: app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt [hunk 3]
```diff
@@ -1,12 +1,269 @@
...
+                    analyzeNode(node.variable, currentScope)
+                    addName(node.variable.name, node.variable, currentScope)
+                }
+
+                is Program.Global.FunctionDefinition -> {
+                    reportIfNameConflict(node.function.name, currentScope)
+
+                    // first add name, then analyze because we can have recursive calls in the body
+                    addName(node.function.name, node.function, currentScope)
+                    analyzeNode(node.function, currentScope)
```
Comment (kamgal00) [inline app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt:146]: There is a small problem here: we allow self-reference in default parameter values: `czynność f(x: Czy = f())`

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt [hunk 4]
```diff
@@ -32,4 +32,14 @@ sealed class Diagnostic {
...
     }
+
+    sealed class NameResolutionErrors() : Diagnostic() {
+        override fun isError() = true
+
+        class UndefinedVariable : NameResolutionErrors()
+        class UndefinedFunction : NameResolutionErrors()
+        class NameConflict : NameResolutionErrors()
+        class VariableIsNotCallable : NameResolutionErrors()
+        class FunctionIsNotVariable : NameResolutionErrors()
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:43]: https://github.com/TCS-Compilers/compiler/pull/53#discussion_r1020980440

**Review by tomekb234 — COMMENTED** [review]
`gradle build` runs ktlint and shows its errors, so you can use this before pushing and not have GitHub reject your commits.

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt [hunk 5]
```diff
@@ -32,4 +32,14 @@ sealed class Diagnostic {
                 append(" Expected symbols: ${expectedSymbols.joinToString()}.")
         }.toString()
     }
+
+    sealed class NameResolutionErrors() : Diagnostic() {
+        override fun isError() = true
+
+        class UndefinedVariable : NameResolutionErrors()
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:39]: Don't you think that we should print some more info, like in parser/lexer? I.e. name of the variable, line, etc.?

File: app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt (hunk 2 above)
Comment (woocash2) [reply app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt:230]: good point

File: app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt (hunk 3 above)
Comment (woocash2) [reply app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt:146]: Right, I think I can move adding names from `FunctionDefinition` to `Function` and from `VariableDefinition` to `Variable`

**Pull request merged**